File: ComputerVision/source/message.py
import json


class Message:
    """Represents a message to be to and from the server"""

    _code: int = None # Identifier (must be positive)
    _data: str = None # Data in JSON string format
    
    def __init__(self, code: int, data: str):

        # Check code validity
        if not type(code) is int or code < 1:
            raise ValueError("Argument 'code' must be a int larger than 0")
        self._code = code

        # Parse JSON and check validity
        try:
            json.loads(data)
        except (json.JSONDecodeError, TypeError):
            raise ValueError("Argument 'data' must be a JSON formatted string")
        self._code = code

        # Parse JSON and check validity
        try:
            json.loads(data)
        except (json.JSONDecodeError, TypeError):
            raise ValueError("Argument 'data' must be a JSON formatted string")
        self._data = data


    def __str__(self):
        return f"Message( code={self._code}, data={self._data} )"


    def to_json(self):
        return json.dumps({"code" : self._code, "data" : json.loads(self._data)})

    @staticmethod
    def from_json(json_msg: str):
        dict_msg = json.loads(json_msg)
        return Message(dict_msg["code"], json.dumps(dict_msg["data"]))



if __name__ == "__main__":
    msg = Message(1, '{"num" : 10}')
    print(Message.from_json(msg.to_json()))
    

File: src/main/java/cv/communication/Message.java
package cv.communication;

import org.json.JSONException;
import org.json.JSONObject;

/** Represents a message to be sent to and from the
 *  Client. **/
public class Message {
    private int code;
    private JSONObject data;

    /**
     * Construct a new message to be sent to the Client, or wrap the message
     * sent from the Client
     * @param code The code representing the message type. Must larger than 0
     * @param data The data within the message ("the message itself"). Can be null
     */
    public Message(int code, JSONObject data) throws IllegalArgumentException {
        if( code < 1 )
            throw new IllegalArgumentException("Message code must be positive");
        this.code = code;

        // Just to ensure that the data property always exists
        if(data == null)
            this.data = new JSONObject();
        else
            this.data = data;
    }

    // For construction of object from JSON
    private Message(){}

    public int getCode() {
        return code;
    }

    public JSONObject getData() {
        return data;
    }


    /**
     * Constructs a Message object from the data of a JSON object.
     * @param jsonObject The JSON object to convert from which must conform to the format: {'code': [int], 'data' : { [data] }}
     * @return A new Message object containing the data within the JSON object
     * @throws JSONException If the message does not conform to the given format
     */
    public static Message fromJSON(JSONObject jsonObject) throws JSONException {
        // Check we don't too many properties
        for(Object name : jsonObject.names())
            if( !name.equals("code") && !name.equals("data") )
                throw new JSONException("Unknown property '" + name + "' in JSON object");

        // Check if code is valid
        int code = jsonObject.getInt("code");
        if( code < 1 )
            throw new JSONException("Property 'code' must be larger than 0");

        // Check that data exsists
        JSONObject data = jsonObject.getJSONObject("data");
        return new Message(code, data);
    }

    public JSONObject toJSON() {
        return new JSONObject(this);
    }

    @Override
    public String toString() {
        return "Message{ " +
                "code=" + code +
                ", data=" + data +
                " }";
    }

}
